File: src/core/errors/result.ts
/**
 * Result<T> — единый контракт возврата для всех модулей.
 *
 * Аналог sealed class из Flutter-версии.
 * Каждый модуль (OCR, CV, ASR, TTS) возвращает Result,
 * Command Processor обрабатывает через discriminated union.
 *
 * @example
 * ```ts
 * const result = await ocrService.recognize(imageBytes);
 * if (result.ok) {
 *   tts.speak(result.data);
 * } else {
 *   logger.error(result.errorCode);
 *   tts.speak(result.userMessage);
 * }
 * ```
 */

export type Result<T> = Success<T> | Failure;

export interface Success<T> {
  readonly ok: true;
  readonly data: T;
}

export interface Failure {
  readonly ok: false;
  readonly errorCode: string;
  readonly userMessage: string;
}

/** Фабрики для удобного создания */
export function success<T>(data: T): Success<T> {
  return {ok: true, data};
}

export function failure(errorCode: string, userMessage: string): Failure {
  return {ok: false, errorCode, userMessage};
}

File: lib/core/errors/result.dart
import 'package:equatable/equatable.dart';

/// Универсальный тип результата для всех модулей (§8.4 ТЗ).
///
/// Каждый модуль (OCR, CV, ASR, TTS) возвращает `Result<T>`,
/// что позволяет Command Processor единообразно обрабатывать
/// успехи и ошибки, логировать и озвучивать fallback-сообщения.
///
/// Пример использования:
/// ```dart
/// final result = await ocrService.recognizeText(image);
/// switch (result) {
///   case Success(:final data):
///     tts.speak(data);
///   case Failure(:final userMessage):
///     tts.speak(userMessage);
/// }
/// ```
sealed class Result<T> extends Equatable {
  const Result();
}

/// Успешный результат с данными типа [T].
class Success<T> extends Result<T> {
  const Success(this.data);

  final T data;

  @override
  List<Object?> get props => [data];
}

/// Результат с ошибкой.
///
/// [errorCode] — машиночитаемый код для логирования (см. [ErrorCode]).
/// [userMessage] — человекочитаемое сообщение для озвучивания через TTS.
class Failure<T> extends Result<T> {
  const Failure({
    required this.errorCode,
    required this.userMessage,
  });

  final String errorCode;
  final String userMessage;

  @override
  List<Object?> get props => [errorCode, userMessage];
}
